src/app/qgsattributetabledialog.cpp

```diff
@@ -890,14 +890,18 @@ void QgsAttributeTableDialog::mActionAddAttribute_triggered()
     if ( mLayer->addAttribute( dialog.field() ) )
     {
       mLayer->endEditCommand();
+
+      if ( mLayer->fields().count() == 1 )
+      {
+        mLayer->setDisplayExpression( dialog.field().name() );
```

Comment:
You should check if it was not defined before no?
Not sure if it can defined on feature id or geometry or constant string.

Reply:
Reworked so set display expression only if a placeholder text is used.